### DISCUSSION TITLE
How can we manage core functionality for plugins?

### DISCUSSION THREAD

**Discussion by cvchaparro** [DevOps-MBSE/AaC#80] [Q&A]
In the task for generating a System Design document #12, there's a requirement for generating a Markdown file containing the document outline. The question then arises, however, about how to separate the core document generation functionality (e.g. assigning certain heading levels to model components, extracting descriptions, etc) from generating the Markdown file (e.g. generating an HTML heading string, etc).

The question that I want us to answer is basically this: What happens when someone wants to generate a System Design document but they want it to be generated as reStructured Text or Org mode or some other format. In addition to a user wanting the file generated in a different file format, I think it makes sense for the infrastructure to be there so that they are able to implement that functionality without the core AaC application/plugins from having to be changed - i.e. think about the Open/Closed principle.

The implementation of this functionality can take a couple approaches.

1. Put the core document generation functionality - i.e. getting the information from the model - into a package and have other packages/modules import the core functionality and generate the document using whatever format is desired.
2. Put the core document generation functionality in the core plugin and have "sub-plugins" that generate the document using whatever format is desired.
3. Put the core document generation functionality in the AaC project (perhaps as just another module) and have plugins that just implement the portion of generating Markdown files, reStructured Text files, etc. using that core functionality.
4. (Similar to 1, minus the extra layer of plugins) Put the core document generation functionality in a plugin and register callbacks (I don't know how we'd do this, yet, honestly) to generate the document using whatever format is desired.

Note: For option 3 above, I think this could make the most sense if we will be re-using the core functionality of getting the information from a model in more than just this plugin (e.g. in the plantuml plugin, any plugin that might be used for requirements tracing, etc).

**Comment by lizzcondrey** [comment]
So, I believe this is a nice set up for providing additional functionality to end users by giving an easy to expand upon core capability. I've not seen or heard much around this topic since joining the group. Was it dropped because of potentially "never ending" scope in that there is always another file format to provide? Was it addressed in a different manner? 

Out of the provided solution options, I think option 3 makes the most sense with the current design of AaC. It would be an additional/expanded API for extracting the information from the model to use as necessary. That the some of this may already be there now with the context API?

**Comment by jondavid-black** [comment]
This is worth more discussion, but suggest we be cautious not to embrace too much flexibility / complexity.  Document formats seem to be like people's children, the are unique and deeply loved.

If we decide to pursue something like this, my initial thoughts would be:
1. Provide some AaC-like way to define a model of your document.  Perhaps this would look similar in some way to what we've got in place for requirements specification allowing for things like headers, footers, title page, sections, subsections, and specific content.  I haven't thought through this enough to recommend a specific Schema for defining documents, but I think it would be an interesting discussion point to explore.
2. Provide some way to specify an output type (i.e. markdown, html, pdf, word).  We would choose to support certain output types and not others.
3. Provide some way to call other plugins and reference their output for inclusion in the document.  Imagine we were to develop a plugin that created a graphic of some sort...perhaps a network architecture.  It would be valuable to capture the output of that command and integrate it into the document.
4. Provide some way to capture "external" artifacts for inclusion in the model.  An example may be test output, pipeline status/artifacts, etc.  We would have to establish constraints of some kind, I'm just not sure what those should be yet.

At the risk of piling on, it may be equally or more valuable to enable the generation of charts.  How many times do we hear "I need a chart for X" when becomes a permanent expectation.  Later we hear something like "it would be so nice to automate this so I don't have to spend my time building charts."  Maybe this is a different thing, but it could be worthy of further discussion.

**Reply by lizzcondrey** [reply]
Puts me in mind of the occasional comments I've heard about a "report plugin" and just report generation needs I've heard on projects, like in your last comment. Agree it merits more discussion because it is something that can spiral in scope quickly and easily. 

I believe I've seen some other discussion thread titles with similar "reporting" themes. I'll review those and see about compiling to one spot as necessary.

**Reply by jondavid-black** [reply]
I can see document generation and reporting becoming a related set of plugins, but I think there could be a number of future plugins that could benefit from a reporting capability.  My recommendation would be to maintain separate plugins for reporting and document generation.  I can imagine a reporting capability that could produce output structured output for use in custom 3rd party scripts/tools.

I think some of the work I'm doing now will be valuable for a reporting plugin in the future.
